samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java

```diff
@@ -161,4 +237,140 @@ void handleExpiredRequestWithHostAffinityEnabled(String processorId, String pref
+   * host on which container is running, container placement action is treated as a restart.
+   *
+   * When host affinity is disabled a move / restart is only allowed on ANY_HOST
+   * When host affinity is enabled move / restart is allowed on specific or ANY_HOST
+   * TODO: SAMZA-2378: Container Placements for Standby containers enabled jobs
+   *
+   * @param processorId logical id of the container 0, 1, 2
+   * @param destinationHost host where container is desired to be moved, acceptable values of this param are any valid
+   *                        hostname or "ANY_HOST"(in this case the request is sent to resource manager for any host)
+   * @param containerAllocator to request physical resources
+   */
+  public ContainerPlacementStatus registerContainerPlacementAction(String processorId, String destinationHost,
+      ContainerAllocator containerAllocator, Optional<Long> requestExpiry) {
+    LOG.info("Received ControlAction request to move or restart container with processor id {} to host {}", processorId, destinationHost);
+    ContainerPlacementStatus actionStatus = checkValidControlAction(processorId, destinationHost);
+    if (actionStatus.status == ContainerPlacementStatus.StatusCode.BAD_REQUEST) {
+      return actionStatus;
+    }
+
+    SamzaResource currentResource = samzaApplicationState.runningProcessors.get(processorId);
+    LOG.info("Processor ID: {} matched a active container with deployment ID: {} running on host: {}", processorId,
+        currentResource.getContainerId(), currentResource.getHost());
+
+    if (destinationHost.equals(ANY_HOST) || !hostAffinityEnabled) {
+      LOG.info("Changing the requested host to {} because either it is requested or host affinity is disabled",
+          ResourceRequestState.ANY_HOST);
+      destinationHost = ANY_HOST;
+    }
+
+    SamzaResourceRequest resourceRequest = containerAllocator.getResourceRequest(processorId, destinationHost);
+    ContainerPlacementMetadata actionMetaData =
+        new ContainerPlacementMetadata(processorId, currentResource.getContainerId(), currentResource.getHost(),
+            destinationHost, actionStatus, requestExpiry.isPresent() ? requestExpiry.get() : DEFAULT_CONTROL_ACTION_EXPIRY);
+
+    // Record the resource request for monitoring
+    actionMetaData.setActionStatus(ContainerPlacementStatus.StatusCode.IN_PROGRESS);
+    actionMetaData.recordResourceRequest(resourceRequest);
+    actions.put(processorId, actionMetaData);
+    // note this also updates state.preferredHost count
+    containerAllocator.issueResourceRequest(resourceRequest);
+    LOG.info("Control action with metadata {} and issued a request for resources in progress", actionMetaData);
+    return actionStatus;
+  }
+
+  public Optional<Long> getActionExpiryTimeout(String processorId) {
+    return this.actions.containsKey(processorId) ? Optional.of(
+        this.actions.get(processorId).getRequestActionExpiryTimeout()) : Optional.empty();
```

Comment:
why do we return `empty` here instead of the default timeout for action defined above? Consider using Duration and also remove `Optional` if possible since there is a default value defined.